storage-blobs-java-v10-tutorial-master/src/main/java/tutorial/Tutorial.java

```diff
@@ -49,109 +38,72 @@ static File createTempFile() throws IOException {
+                    try {
+                        AsynchronousFileChannel channel = AsynchronousFileChannel.open(Paths.get(sourceFile.getPath()), StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+                        transformer.getValue().subscribe(bufferValue -> {
+                            channel.write(bufferValue, 0);
+                            System.out.println("The blob was downloaded to " + sourceFile.getAbsolutePath());
+                        });
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                    transformer.getValue().subscribe();
+                });
     }
 
-    static void listBlobs(ContainerURL containerURL) {
-        // Each ContainerURL.listBlobsFlatSegment call return up to maxResults (maxResults=10 passed into ListBlobOptions below).
-        // To list all Blobs, we are creating a helper static method called listAllBlobs,
-    	// and calling it after the initial listBlobsFlatSegment call
+    static void listBlobs(BlobContainerAsyncClient blobContainerAsyncClient) {
+        // Each BlobContainerAsyncClient.listBlobs call return up to maxResultsPerPage (maxResults=10 passed into ListBlobOptions below).
         ListBlobsOptions options = new ListBlobsOptions();
-        options.withMaxResults(10);
-
-        containerURL.listBlobsFlatSegment(null, options, null).flatMap(containerListBlobFlatSegmentResponse ->
-            listAllBlobs(containerURL, containerListBlobFlatSegmentResponse))
-            .subscribe(response-> {
-                System.out.println("Completed list blobs request.");
-                System.out.println(response.statusCode());
-            });
-    }
-
-    private static Single <ContainerListBlobFlatSegmentResponse> listAllBlobs(ContainerURL url, ContainerListBlobFlatSegmentResponse response) {
-        // Process the blobs returned in this result segment (if the segment is empty, blobs() will be null.
-        if (response.body().segment() != null) {
-            for (BlobItem b : response.body().segment().blobItems()) {
-                String output = "Blob name: " + b.name();
-                if (b.snapshot() != null) {
-                    output += ", Snapshot: " + b.snapshot();
-                }
-                System.out.println(output);
+        options.setMaxResultsPerPage(10);
+        blobContainerAsyncClient.listBlobs(options).count().subscribe(count -> {
+            if (count > 0) {
+                blobContainerAsyncClient.listBlobs(options).subscribe(response -> {
+                    String output = "Blob name: " + response.getName();
+                    if (response.getSnapshot() != null) {
+                        output += ", Snapshot: " + response.getSnapshot();
+                    }
+                    System.out.println(output);
+                });
+            } else {
+                System.out.println("There are no more blobs to list off.");
```

Comment:
Instead of listing twice here this would be better

```suggestion
        blobContainerAsyncClient.listBlobs(options).subscribe(response -> {
                String output = "Blob name: " + response.getName();
                if (response.getSnapshot() != null) {
                    output += ", Snapshot: " + response.getSnapshot();
                }
                System.out.println(output);
            },
            // Error consumer for if there is an error during blob listing, null is a no-op.
            null,
            // Completion Runnable for when listing completes
            () -> System.out.println("There are no more blobs to list."));
```

Reply:
Fixed in the new version